### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by ichsansandy**
In these project I implement

- Install CanCanCan in your project.
- Add a role column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
	- Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
	-  Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -14,6 +14,15 @@ def new
     @post = Post.new
   end
```
Comment (topeogunleye) [inline app/controllers/posts_controller.rb:19]: Hello and great work on the `PostsController`! 🎉 

You’ve done an impressive job structuring your methods and utilizing Rails features. However, after reviewing the code and the project requirements related to CanCanCan authorization, there is a key area that we need to address to ensure alignment with the specified needs:

#### **CanCanCan Authorization Usage:**
- **Project Requirement:** Implement CanCanCan for Post authorization using `load_and_authorize_resource` at the beginning of the controller.
- **Your Implementation:** Manually applied authorization check using `authorize!` method in the `destroy` action.
  
  Suggested change:
  ```ruby
  class PostsController < ApplicationController
    load_and_authorize_resource
    
    ...
  end
  ```
  Implementing `load_and_authorize_resource` will automate the loading and authorization of resources in all actions, enhancing DRY (Don’t Repeat Yourself) practices.

### Revised Code Example (Partial):
```ruby
class PostsController < ApplicationController
  load_and_authorize_resource :user
  load_and_authorize_resource :post, through: :user
  
  before_action :set_user, only: [:index, :show]

  ...
  
  def destroy
    @author = @post.author
    @author.decrement!(:post_counter)
    @post.destroy!
    redirect_to user_posts_path(@user), notice: 'Post was successfully deleted.'
  end
  
  private
  
  def set_user
    @user = User.find(params[:user_id])
  end
end
```
   
### Wrapping Up:
Your coding skills are clearly evident and the logic implemented in the controller is solid! The key takeaway here is to maximize the use of available tools and methodologies (like CanCanCan) to automate repetitive tasks, enhancing code maintainability and readability. 🚀

Keep up the awesome work, and if you have any questions or would like further clarification on the feedback, please don’t hesitate to reach out.

Happy coding! 🧑‍💻🔥

File: app/models/user.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable, :confirmable
+
```
Comment (topeogunleye) [inline app/models/user.rb:6]: [OPTIONAL]
#### Areas for Improvement and Suggested Actions:

1. **Refinement in User Model:**
   - **Observation:** It seems the `admin?` method, which checks if a user is an admin, is not defined within the `User` model.
   - **Suggested Action:** Creating a method like this is a good practice for encapsulating the logic that determines user roles, ensuring cleaner and DRYer code.

```ruby
# app/models/user.rb
...
def admin?
  role == 'admin'
end
```
   
   Implementing this method could streamline your ability class and any future instances where admin checks are needed, improving code maintainability and readability.

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi,  @ichsansandy, @kessie2862 

Good job so far!, Hello and thank you for your diligent efforts on the project! Your work with CanCanCan and role-based authorizations in Rails is taking a great shape.
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪🌟📚

1. **Embedding CanCanCan:**
   - **Fantastic Job with integrating CanCanCan!** Including `gem 'cancancan'` in your Gemfile is crucial for leveraging its functionalities, and you've nailed this step!

2. **Constructive Ability Class:**
   - **A Round of Applause for developing a structured `Ability` class!** 🎉 The rules you've set for `Post` and `Comment` models are clear and well articulated. This demonstration of managing user-specific and admin permissions is crucial and implemented effectively.

3. **Role Column Implementation:**
   - **Spot On with the `role` column in the users table!** 🎯 Ensuring this column is in your migration files underpins the role-based authorization strategy and you've ensured it’s appropriately configured.

Happy Coding! 🎉🧑‍💻

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Please kindly check my review comments.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by AbayomiOlaoye — APPROVED** [review]
## STATUS: APPROVED :green_circle:  

Hi @Team  :wave: :wave:, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

applause

## Highlights
- Cleanly written code :white_check_mark: 
- App runs well and functions as expected ✅ 
- Nicely implemented UI styles ✅ 
- Great teamwork ✅ 
- Good response to feedback :heavy_check_mark: 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- [ ] Kindly give equal attention to your commit messages as they depict your commitment to best practices and professionalism. As a principle, it is more correct to always ensure that your first letter in any sentence is capitalized.

image


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**